README.md
# nepninja
Reuse the beautiful animated widgets with less effort

## Sample
![](https://github.com/khadkarajesh/nepninja/blob/master/assets/images/circular_animated_button.gif)

## Usage

```dart

 CircularProgressButton(
            height: 55,
            width: 200,
            borderRadius: 30,
            backgroundColor: Colors.lightBlueAccent,
            fadeDurationInMilliSecond: 400,
            text: "Sign In",
            fontSize: 30,
            progressIndicatorColor: Colors.pinkAccent,
            onTap: (reset) {
              makeRequest(reset);
            },
          )

 # reset state once validation error or network fails
 void makeRequest(Function reset) {
    Future.delayed(Duration(seconds: 3), () {
      reset();
    });
  }

```

## Article
You can read further more detail about implementation over this [article](https://medium.com/@rajesh_khadka/flutter-animation-challenge-1-5ce9bed65161).


CHANGELOG.md
## [0.0.1] - 2020-04-26

* Added the circular progress button animation

## [0.0.2] - 2020-04-26
* Update readme.md

## [0.0.3] - 2020-04-26
* Remove unused description on lib

## [0.0.4] - 2020-04-26
* Fix issue of import on lib


test/nepninja_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:nepninja/nepninja.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/circular_progress_button.dart
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';

// ignore: must_be_immutable
class CircularProgressButton extends StatefulWidget {
  double width;
  double height;
  int borderRadius;
  Color backgroundColor;
  int fadeDurationInMilliSecond;
  String text;
  Color progressIndicatorColor;
  double fontSize;
  final void Function(Function reset) onTap;

  CircularProgressButton({
    this.height = 55.0,
    this.width = 200.0,
    this.borderRadius = 30,
    this.backgroundColor = Colors.lightBlueAccent,
    this.fadeDurationInMilliSecond = 700,
    this.text = "Click",
    this.progressIndicatorColor = Colors.pinkAccent,
    this.fontSize = 16.0,
    this.onTap,
  });

  @override
  State<StatefulWidget> createState() {
    return _StateAnimationButton(
      height: this.height,
      width: this.width,
      borderRadius: this.borderRadius,
      backgroundColor: this.backgroundColor,
      fadeDurationInMilliSecond: this.fadeDurationInMilliSecond,
      text: this.text,
      progressIndicatorColor: this.progressIndicatorColor,
      fontSize: this.fontSize,
      onTap: onTap,
    );
  }
}

class _StateAnimationButton extends State<CircularProgressButton> {
  bool squeezed = false;
  double _opacity = 1.0;
  static const double SQUEEZED_BORDER_RADIUS = 70.0;

  final double width;
  final double height;
  final int borderRadius;
  final Color backgroundColor;
  final int fadeDurationInMilliSecond;
  final String text;
  final Color progressIndicatorColor;
  final double fontSize;
  final void Function(Function reset) onTap;

  _StateAnimationButton({
    @required this.height,
    @required this.width,
    @required this.borderRadius,
    @required this.backgroundColor,
    @required this.fadeDurationInMilliSecond,
    @required this.text,
    @required this.progressIndicatorColor,
    @required this.fontSize,
    @required this.onTap,
  });

  void reset() {
    setState(() {
      squeezed = !squeezed;
      _opacity = _opacity == 1.0 ? 0.0 : 1.0;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        InkWell(
          onTap: () {
            setState(() {
              squeezed = !squeezed;
              _opacity = _opacity == 1.0 ? 0.0 : 1.0;
              onTap(reset);
            });
          },
          child: AnimatedContainer(
            width: squeezed ? height : width,
            height: height,
            duration: Duration(milliseconds: fadeDurationInMilliSecond),
            curve: Curves.fastOutSlowIn,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(
                  squeezed ? SQUEEZED_BORDER_RADIUS : borderRadius.toDouble()),
              color: backgroundColor,
            ),
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                AnimatedOpacity(
                  child: Text(
                    "${this.text}",
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: fontSize < 18.0 ? fontSize : 18.0,
                    ),
                  ),
                  opacity: _opacity,
                  duration: Duration(seconds: 1),
                ),
              ],
            ),
          ),
        ),
        InkWell(
          onTap: () {
            setState(() {
              squeezed = !squeezed;
              _opacity = _opacity == 1.0 ? 0.0 : 1.0;
              onTap(reset);
            });
          },
          child: AnimatedContainer(
            width: squeezed ? height : width,
            height: height,
            duration: Duration(milliseconds: fadeDurationInMilliSecond),
            curve: Curves.fastOutSlowIn,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(
                  squeezed ? SQUEEZED_BORDER_RADIUS : borderRadius.toDouble()),
            ),
            child: AnimatedOpacity(
              child: Padding(
                padding: EdgeInsets.all(1),
                child: CircularProgressIndicator(
                    backgroundColor: backgroundColor,
                    valueColor: AlwaysStoppedAnimation<Color>(
                        squeezed ? progressIndicatorColor : backgroundColor)),
              ),
              opacity: _opacity == 0.0 ? 1.0 : 0.0,
              duration: Duration(milliseconds: fadeDurationInMilliSecond),
            ),
          ),
        ),
      ],
    );
  }
}


example/src/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:nepninja/src/circular_progress_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Nepninja Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Login(),
    );
  }
}

class Login extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _StateLogin();
  }
}

class _StateLogin extends State<Login> {
  @override
  Widget build(BuildContext context) {
    return Container(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          CircularProgressButton(
            height: 55,
            width: 200,
            borderRadius: 30,
            backgroundColor: Colors.lightBlueAccent,
            fadeDurationInMilliSecond: 400,
            text: "Sign In",
            fontSize: 30,
            progressIndicatorColor: Colors.pinkAccent,
            onTap: (reset) {
              makeRequest(reset);
            },
          )
        ],
      ),
    );
  }

  void makeRequest(Function reset) {
    Future.delayed(Duration(seconds: 3), () {
      reset();
    });
  }
}
